File: src/android/JumioMobileSDK.java
/*
 * Copyright 2017 Jumio Corporation
 * All rights reserved
 */

package com.jumio.mobilesdk;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.util.Log;

import com.jumio.defaultui.JumioActivity;
import com.jumio.sdk.JumioSDK;
import com.jumio.sdk.credentials.JumioCredentialInfo;
import com.jumio.sdk.result.JumioCredentialResult;
import com.jumio.sdk.result.JumioFaceResult;
import com.jumio.sdk.result.JumioIDResult;
import com.jumio.sdk.result.JumioResult;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class JumioMobileSDK extends CordovaPlugin {

    private static String TAG = "JumioMobileSDK";
    private static final int PERMISSION_REQUEST_CODE = 301;
    private final static int REQUEST_CODE = 101;

    private static final String ACTION_INIT = "initialize";
    private static final String ACTION_START = "start";

    private CallbackContext callbackContext;

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) {
        PluginResult result = null;
        this.callbackContext = callbackContext;

        if (action.equals(ACTION_INIT)) {
            initialize(args);
            result = new PluginResult(Status.NO_RESULT);
            result.setKeepCallback(false);
            return true;
        } else if (action.equals(ACTION_START)) {
            start();
            result = new PluginResult(Status.NO_RESULT);
            result.setKeepCallback(true);
            return true;
        } else {
            result = new PluginResult(Status.INVALID_ACTION);
            callbackContext.error("Invalid Action");
            return false;
        }
    }

    private void initialize(JSONArray data) {
        if (!JumioSDK.isSupportedPlatform(cordova.getActivity())) {
            showErrorMessage("This platform is not supported.");
            return;
        }

        try {
            if (data.isNull(0) || data.isNull(1)) {
                showErrorMessage("Missing required parameters one-time session authorization token or dataCenter.");
                return;
            }

            final String authorizationToken = data.getString(0);
            final String dataCenter = data.getString(1);

            final Intent intent = new Intent(cordova.getActivity(), JumioActivity.class);
            intent.putExtra(JumioActivity.EXTRA_TOKEN, authorizationToken);
            intent.putExtra(JumioActivity.EXTRA_DATACENTER, dataCenter);

            //The following intent extra can be used to customize the Theme of Default UI
//            intent.putExtra(JumioActivity.EXTRA_CUSTOM_THEME, R.style.AppThemeCustomJumio);

            cordova.getActivity().startActivityForResult(intent, REQUEST_CODE);

        } catch (JSONException e) {
            showErrorMessage("Invalid parameters: " + e.getLocalizedMessage());
        }
    }

    private void start() {
        Runnable runnable = () -> {
            try {
                checkPermissionsAndStart();
            } catch (Exception e) {
                showErrorMessage("Error starting the Jumio SDK: " + e.getLocalizedMessage());
            }
        };

        this.cordova.setActivityResultCallback(this);
        this.cordova.getActivity().runOnUiThread(runnable);
    }

    // Permissions
    private void checkPermissionsAndStart() {
        if (!JumioSDK.hasAllRequiredPermissions(cordova.getActivity().getApplicationContext())) {
            //Acquire missing permissions.
            String[] mp = JumioSDK.getMissingPermissions(cordova.getActivity().getApplicationContext());

            cordova.requestPermissions(this, PERMISSION_REQUEST_CODE, mp);
        }
    }

    @Override
    public void onRequestPermissionResult(int requestCode, String[] permissions, int[] grantResults) throws
            JSONException {
        boolean allGranted = true;
        for (int grantResult : grantResults) {
            if (grantResult != PackageManager.PERMISSION_GRANTED) {
                allGranted = false;
                break;
            }
        }

        if (!allGranted) {
            showErrorMessage("You need to grant all required permissions to continue");
            super.onRequestPermissionResult(requestCode, permissions, grantResults);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_CODE) {
            if (data == null) {
                return;
            }

            final JumioResult jumioResult = (JumioResult) data.getSerializableExtra(JumioActivity.EXTRA_RESULT);

            if (jumioResult != null && jumioResult.isSuccess()) {
                sendScanResult(jumioResult);
            } else {
                sendCancelResult(jumioResult);
            }
        }
    }

    private void sendScanResult(final JumioResult jumioResult) {
        final String accountId = jumioResult.getAccountId();
        final List<JumioCredentialInfo> credentialInfoList = jumioResult.getCredentialInfos();

        final JSONObject result = new JSONObject();
        final ArrayList<JSONObject> credentialsArray = new ArrayList<>();


        if (credentialInfoList != null) {
            try {
                if (accountId != null) {
                    result.put("accountId", accountId);
                }
            } catch (JSONException e) {
                showErrorMessage("Result could not be sent: " + e.getLocalizedMessage());
            }

            for (JumioCredentialInfo credentialInfo : credentialInfoList) {
                final JumioCredentialResult jumioCredentialResult = jumioResult.getResult(credentialInfo);
                final JSONObject credentialMap = new JSONObject();
                try {
                    credentialMap.put("credentialId", credentialInfo.getId());
                    credentialMap.put("credentialCategory", credentialInfo.getCategory().toString());


                    if (jumioCredentialResult instanceof JumioIDResult) {
                        final JumioIDResult idResult = (JumioIDResult) jumioCredentialResult;
                        credentialMap.put("selectedCountry", idResult.getCountry());
                        credentialMap.put("selectedDocumentType", idResult.getIdType());
                        credentialMap.put("idNumber", idResult.getDocumentNumber());
                        credentialMap.put("personalNumber", idResult.getPersonalNumber());
                        credentialMap.put("issuingDate", idResult.getIssuingDate());
                        credentialMap.put("expiryDate", idResult.getExpiryDate());
                        credentialMap.put("issuingCountry", idResult.getIssuingCountry());
                        credentialMap.put("lastName", idResult.getLastName());
                        credentialMap.put("firstName", idResult.getFirstName());
                        credentialMap.put("gender", idResult.getGender());
                        credentialMap.put("nationality", idResult.getNationality());
                        credentialMap.put("dateOfBirth", idResult.getDateOfBirth());

                        credentialMap.put("addressLine", idResult.getAddress());
                        credentialMap.put("city", idResult.getCity());
                        credentialMap.put("subdivision", idResult.getSubdivision());
                        credentialMap.put("postCode", idResult.getPostalCode());
                        credentialMap.put("placeOfBirth", idResult.getPlaceOfBirth());

                        credentialMap.put("mrzLine1", idResult.getMrzLine1());
                        credentialMap.put("mrzLine2", idResult.getMrzLine2());
                        credentialMap.put("mrzLine3", idResult.getMrzLine3());
                    } else if (jumioCredentialResult instanceof JumioFaceResult) {
                        //lowercase
                        credentialMap.put("passed", String.valueOf(((JumioFaceResult) jumioCredentialResult).getPassed()));
                    }
                } catch (JSONException e) {
                    showErrorMessage("Result could not be sent: " + e.getLocalizedMessage());
                }
                credentialsArray.add(credentialMap);
            }
            try {
                result.put("credentials", credentialsArray);
            } catch (JSONException e) {
                showErrorMessage("Result could not be sent: " + e.getLocalizedMessage());
            }
        }
        callbackContext.success(result);
    }

    private void sendCancelResult(final JumioResult jumioResult) {
        if (jumioResult != null && jumioResult.getError() != null) {
            String errorMessage = jumioResult.getError().getMessage();
            String errorCode = jumioResult.getError().getCode();
            sendErrorObject(errorCode, errorMessage);
        } else {
            showErrorMessage("There was a problem extracting the scan result");
        }
    }

    private void showErrorMessage(String msg) {
        Log.e(TAG, msg);
        try {
            JSONObject errorResult = new JSONObject();
            errorResult.put("errorMessage", msg != null ? msg : "");
            if (callbackContext != null) {
                callbackContext.error(errorResult);
            }
        } catch (JSONException | NullPointerException e) {
            Log.e(TAG, e.getLocalizedMessage());
        }
    }

    private void sendErrorObject(String errorCode, String errorMsg) {
        try {
            JSONObject errorResult = new JSONObject();
            errorResult.put("errorCode", errorMsg != null ? errorCode : "");
            errorResult.put("errorMessage", errorMsg != null ? errorMsg : "");

            if(callbackContext != null) {
                callbackContext.error(errorResult);
            }
        } catch (JSONException | NullPointerException e) {
            showErrorMessage("Result could not be sent: " + e.getLocalizedMessage());
        }
    }
}

File: src/android/JumioMobileSDK.kt
package com.jumio.mobilesdk

import android.content.Intent
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.os.Build
import android.util.Log
import com.jumio.defaultui.JumioActivity
import com.jumio.sdk.JumioSDK
import com.jumio.sdk.credentials.JumioCredentialCategory.FACE
import com.jumio.sdk.credentials.JumioCredentialCategory.ID
import com.jumio.sdk.enums.JumioDataCenter
import com.jumio.sdk.preload.JumioPreloadCallback
import com.jumio.sdk.preload.JumioPreloader
import com.jumio.sdk.result.JumioIDResult
import com.jumio.sdk.result.JumioResult
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.PluginResult
import org.apache.cordova.PluginResult.Status.INVALID_ACTION
import org.apache.cordova.PluginResult.Status.NO_RESULT
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class JumioMobileSDK : CordovaPlugin(), JumioPreloadCallback {
    companion object {
        private const val TAG = "JumioMobileSDK"
        private const val PERMISSION_REQUEST_CODE = 301
        private const val REQUEST_CODE = 101
        private const val ACTION_INIT = "initialize"
        private const val ACTION_START = "start"
        private const val ACTION_SET_PRELOADER_FINISHED_BLOCK = "setPreloaderFinishedBlock"
        private const val ACTION_PRELOAD_IF_NEEDED = "preloadIfNeeded"
        private const val ACTION_GET_CACHED_RESULT = "getCachedResult"
        var pendingResult: JSONObject? = null
        var pendingError: JSONObject? = null
    }

    private var callbackContext: CallbackContext? = null
    private var preloaderFinishedCallback: (() -> Unit)? = null
    private var authorizationToken: String? = null
    private var dataCenter: String? = null

    override fun execute(action: String, args: JSONArray, callbackContext: CallbackContext): Boolean {
        if (action == ACTION_START || action == ACTION_SET_PRELOADER_FINISHED_BLOCK || action == ACTION_PRELOAD_IF_NEEDED) {
            this.callbackContext = callbackContext
        }

        val result: PluginResult?
        return when (action) {
            ACTION_INIT -> {
                initialize(args, callbackContext)
                result = PluginResult(NO_RESULT)
                result.keepCallback = false
                true
            }
            ACTION_START -> {
                start()
                result = PluginResult(NO_RESULT)
                result.keepCallback = true
                true
            }
            ACTION_SET_PRELOADER_FINISHED_BLOCK -> {
                setPreloaderFinishedBlock {
                    callbackContext.success()
                }
                result = PluginResult(NO_RESULT)
                result.keepCallback = true
                true
            }
            ACTION_PRELOAD_IF_NEEDED -> {
                preloadIfNeeded()
                result = PluginResult(NO_RESULT)
                result.keepCallback = true
                true
            }
            ACTION_GET_CACHED_RESULT -> {
                checkAndSendCachedResult(callbackContext)
                result = PluginResult(NO_RESULT)
                result.keepCallback = false
                true
            }
            else -> {
                result = PluginResult(INVALID_ACTION)
                callbackContext.error("Invalid Action")
                false
            }
        }
    }

    @Throws(JSONException::class)
    override fun onRequestPermissionResult(requestCode: Int, permissions: Array<String?>?, grantResults: IntArray) {
        if (requestCode == PERMISSION_REQUEST_CODE) {
            val hasMissingPermission = grantResults.any { it != PERMISSION_GRANTED }

            if (hasMissingPermission) {
                showErrorMessage("You need to grant all required permissions to continue")
            } else {
                initSdk()
            }
        } else {
            super.onRequestPermissionResult(requestCode, permissions, grantResults)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_CODE) {
            data?.let {
                val jumioResult = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    it.getSerializableExtra(JumioActivity.EXTRA_RESULT, JumioResult::class.java)
                } else {
                    @Suppress("DEPRECATION")
                    it.getSerializableExtra(JumioActivity.EXTRA_RESULT) as JumioResult?
                }

                if (jumioResult?.isSuccess == true) sendScanResult(jumioResult) else sendCancelResult(jumioResult)
            }
        }
    }

    private fun initialize(data: JSONArray, callbackContext: CallbackContext) {
        val authorizationToken = data.getString(0)
        val dataCenter = data.getString(1)
        val jumioDataCenter = getJumioDataCenter(dataCenter)

        when {
            !JumioSDK.isSupportedPlatform(cordova.activity) -> {
                callbackContext.error("This platform is not supported.")
                return
            }
            authorizationToken.isNullOrBlank() -> {
                callbackContext.error("Missing required parameters one-time session authorization token.")
            }
            jumioDataCenter == null -> {
                callbackContext.error("Invalid Datacenter value.")
            }
            else -> {
                this.authorizationToken = authorizationToken
                this.dataCenter = dataCenter
                callbackContext.success()
            }
        }
    }

    private fun initSdk() {
        val intent = Intent(cordova.activity, JumioActivity::class.java).apply {
            putExtra(JumioActivity.EXTRA_TOKEN, authorizationToken)
            putExtra(JumioActivity.EXTRA_DATACENTER, dataCenter)

            //The following intent extra can be used to customize the Theme of Default UI
            putExtra(JumioActivity.EXTRA_CUSTOM_THEME, cordova.activity.applicationContext.resources.getIdentifier("AppThemeCustomJumio", "style", cordova.activity.applicationContext.packageName))
        }

        cordova.activity.startActivityForResult(intent, REQUEST_CODE)
    }

    private fun start() {
        val runnable = Runnable {
            try {
                checkPermissionsAndStart()
            } catch (e: Exception) {
                showErrorMessage("Error starting the Jumio SDK: " + e.localizedMessage)
            }
        }
        cordova.setActivityResultCallback(this)
        cordova.activity.runOnUiThread(runnable)
    }

    private fun setPreloaderFinishedBlock(completion: (() -> Unit)) {
        with(JumioPreloader) {
            init(cordova.activity)
            setCallback(this@JumioMobileSDK)
        }
        preloaderFinishedCallback = completion
    }

    private fun preloadIfNeeded() {
        with(JumioPreloader) {
            preloadIfNeeded()
        }
    }

    override fun preloadFinished() {
        preloaderFinishedCallback?.invoke()
    }

    // Permissions
    private fun checkPermissionsAndStart() {
        if (!JumioSDK.hasAllRequiredPermissions(cordova.activity.applicationContext)) {
            val permissions = JumioSDK.getMissingPermissions(cordova.activity.applicationContext)
            cordova.requestPermissions(this, PERMISSION_REQUEST_CODE, permissions)
        } else {
            initSdk()
        }
    }

    private fun sendScanResult(jumioResult: JumioResult?) {
        val accountId = jumioResult?.accountId
        val workflowId = jumioResult?.workflowExecutionId
        val credentialInfoList = jumioResult?.credentialInfos

        val result = JSONObject()
        val credentialsArray = ArrayList<JSONObject>()

        try {
            credentialInfoList?.let {
                accountId?.let { result.put("accountId", it) }
                workflowId?.let { result.put("workflowId", it) }

                credentialInfoList.forEach {
                    val credentialMap = JSONObject()
                    credentialMap.put("credentialId", it.id)
                    credentialMap.put("credentialCategory", it.category.toString())

                    if (it.category == ID) {
                        val idResult = jumioResult.getIDResult(it)
                        idResult?.let { handleIdResult(idResult, credentialMap) }
                    } else if (it.category == FACE) {
                        val faceResult = jumioResult.getFaceResult(it)
                        faceResult?.passed?.let { passed -> credentialMap.put("passed", passed.toString()) }
                    }
                    credentialsArray.add(credentialMap)
                }
                result.put("credentials", credentialsArray)
            }
        } catch (e: JSONException) {
            showErrorMessage("Result could not be sent: " + e.localizedMessage)
            return
        }

        if (this.callbackContext != null) {
            this.callbackContext?.success(result)
        } else {
            pendingResult = result
            pendingError = null
        }
        this.callbackContext = null
    }
    private fun handleIdResult(idResult: JumioIDResult, credentialMap: JSONObject) =
        with(idResult) {
            credentialMap.apply {
                country?.let { put("selectedCountry", it) }
                idType?.let { put("selectedDocumentType", it) }
                idSubType?.let { put("selectedDocumentSubType", it) }
                documentNumber?.let { put("idNumber", it) }
                personalNumber?.let { put("personalNumber", it) }
                issuingDate?.let { put("issuingDate", it) }
                expiryDate?.let { put("expiryDate", it) }
                issuingCountry?.let { put("issuingCountry", it) }
                lastName?.let { put("lastName", it) }
                firstName?.let { put("firstName", it) }
                gender?.let { put("gender", it) }
                nationality?.let { put("nationality", it) }
                dateOfBirth?.let { put("dateOfBirth", it) }
                address?.let { put("addressLine", it) }
                city?.let { put("city", it) }
                subdivision?.let { put("subdivision", it) }
                postalCode?.let { put("postCode", it) }
                placeOfBirth?.let { put("placeOfBirth", it) }
                mrzLine1?.let { put("mrzLine1", it) }
                mrzLine2?.let { put("mrzLine2", it) }
                mrzLine3?.let { put("mrzLine3", it) }
            }
        }


    private fun sendCancelResult(jumioResult: JumioResult?) {
        jumioResult?.error?.let {
            sendErrorObject(it.code, it.message)
        } ?: showErrorMessage("There was a problem extracting the scan result")
    }

    private fun showErrorMessage(msg: String?) {
        Log.e(TAG, msg ?: "")
        sendErrorObject("", msg ?: "Unknown error")
    }

    private fun sendErrorObject(errorCode: String, errorMsg: String?) {
        try {
            val errorResult = JSONObject().apply {
                put("errorCode", if (errorMsg != null) errorCode else "")
                put("errorMessage", errorMsg ?: "")
            }

            if (this.callbackContext != null) {
                this.callbackContext?.error(errorResult)
            } else {
                pendingError = errorResult
                pendingResult = null
            }
            this.callbackContext = null
        } catch (e: JSONException) {
            Log.e(TAG, "Result could not be sent: " + e.localizedMessage)
        } catch (e: NullPointerException) {
            Log.e(TAG, "Result could not be sent: " + e.localizedMessage)
        }
    }

    private fun checkAndSendCachedResult(callbackContext: CallbackContext?) {
        pendingResult?.let {
            callbackContext?.success(it)
            pendingResult = null
            pendingError = null
            return
        }
        pendingError?.let {
            callbackContext?.error(it)
            pendingResult = null
            pendingError = null
            return
        }
    }

    private fun getJumioDataCenter(dataCenter: String) = try {
        JumioDataCenter.valueOf(dataCenter)
    } catch (_: IllegalArgumentException) {
        null
    }
}